gobblin-data-management/src/main/java/org/apache/gobblin/data/management/source/LoopingDatasetFinderSource.java

```diff
@@ -85,53 +88,82 @@ public WorkUnitStream getWorkunitStream(SourceState state) {
       int maxWorkUnits = state.getPropAsInt(MAX_WORK_UNITS_PER_RUN_KEY, MAX_WORK_UNITS_PER_RUN);
 
       List<WorkUnitState> previousWorkUnitStates = state.getPreviousWorkUnitStates();
-      Optional<WorkUnitState> maxWorkUnit;
-      try {
-        maxWorkUnit = previousWorkUnitStates.stream().reduce((wu1, wu2) -> {
-          int wu1Ordinal = wu1.getPropAsInt(WORK_UNIT_ORDINAL);
-          int wu2Ordinal = wu2.getPropAsInt(WORK_UNIT_ORDINAL);
-          return wu1Ordinal > wu2Ordinal ? wu1 : wu2;
-        });
-      } catch (NumberFormatException nfe) {
-        throw new RuntimeException("Work units in state store are corrupted! Missing or malformed " + WORK_UNIT_ORDINAL);
+      int tempWorkUnitOrdinal = 0;
+      Optional<WorkUnitState> optionalMaxWorkUnit = Optional.empty();
+      List<WorkUnit> failedPreviousWorkUnits = new ArrayList<>();
+      int maxRetryTimes = state.contains(ConfigurationKeys.MAX_TASK_CROSSEXECUTION_TASK_RETRIES_KEY) ? state
+          .getPropAsInt(ConfigurationKeys.MAX_TASK_CROSSEXECUTION_TASK_RETRIES_KEY)
+          : ConfigurationKeys.DEFAULT_MAX_TASK_RETRIES;
+
+      // Searching for workunit with maximum ordinal meanwhile collect all failed workunit in previous run.
+      // These two processes should not influent each other.
+
+      // Workunits that exceed max retry limit will not be added into failedPreviousWorkUnits
+      // and rollovered into next quota-available execution.
+      for (WorkUnitState workUnitState : previousWorkUnitStates) {
+        if ((workUnitState.getWorkingState() == WorkUnitState.WorkingState.FAILED
+            || workUnitState.getWorkingState() == WorkUnitState.WorkingState.ROLLOVERED) && (
+            !workUnitState.contains(ConfigurationKeys.TASK_RETRIES_KEY)
+                || workUnitState.getPropAsInt(ConfigurationKeys.TASK_RETRIES_KEY) < maxRetryTimes + 1)) {
+          log.info("Dataset " + workUnitState.getProp(DATASET_URN) + " is failed previously, retrying...");
+          WorkUnit retryWorkUnit = new WorkUnit(workUnitState.getWorkunit());
+          failedPreviousWorkUnits.add(retryWorkUnit);
+        }
```

Comment:
And if the max retries exceeds the maxRetryTimes, we just silently drop the workunit and never try processing it again? A few questions:

* Before your change here, would failed workunits just be retried forever in all subsequent executions?
* Does Gobblin have a concept of "permanently failed" workunits that are written to a separate location and never retried in subsequent executions?

Reply:
max retries exceeds the limit, a event will be sent and it will result in a failed workunit therefore never being retried automatically. But users/admins should be aware of this by our notification mechanism which I believe has not been fully discussed yet @ibuenros . 

For your questions: 
- No. Current retry model will retry the failed workunit in current execution until reaching the maximum retry limit. 
- No. I think we should discuss the notification mechanism about such kind of failure.

Reply:
Here, I think we should add least `log.warn()` about the workunits we are skipping because they've exceeded the max retries.